src/io/io.c

```diff
@@ -157,10 +177,13 @@ MVMObject * MVM_io_read_bytes_async(MVMThreadContext *tc, MVMObject *oshandle, M
                                     MVMObject *schedulee, MVMObject *buf_type, MVMObject *async_type) {
     MVMOSHandle *handle = verify_is_handle(tc, oshandle, "read bytes asynchronously");
     if (handle->body.ops->async_readable) {
-        uv_mutex_t *mutex = acquire_mutex(tc, handle);
-        MVMObject *result = (MVMObject *)handle->body.ops->async_readable->read_bytes(tc,
-            handle, queue, schedulee, buf_type, async_type);
-        release_mutex(tc, mutex);
+        MVMObject *result;
+        MVMROOT(tc, handle, {
+            uv_mutex_t *mutex = acquire_mutex(tc, handle);
+            result = (MVMObject *)handle->body.ops->async_readable->read_bytes(tc,
+                handle, queue, schedulee, buf_type, async_type);
```

Comment:
queue, schedulee, buf_type and async_type must also be MVMROOTed.
All pointers to garbage collected objects (like MVMObject and subtypes) that you access after a GC triggering function (acquire_mutex here) must be rooted. Doesn't matter if you use them within the MVMROOT block or after it. The MVMROOT block is the part where GC may be triggered and thus the pointer may have to be updated.

Reply:
Thanks for being explicit about the what and why. Will attempt to address later today.